plugins/win-mf/mf-aac.cpp

```diff
@@ -0,0 +1,161 @@
+#include <obs-module.h>
+
+#include <memory>
+
+#include "mf-aac-encoder.hpp"
+
+#include <VersionHelpers.h>
+
+using namespace MFAAC;
+
+namespace {
+const char *MFAAC_GetName(void *)
+{
+	return obs_module_text("MFAACEnc");
+}
+
+obs_properties_t *MFAAC_GetProperties(void *)
+{
+	auto *props = obs_properties_create();
+
+	obs_properties_add_int(props, "bitrate", obs_module_text("Bitrate"), 96, 192, 32);
+
+	return props;
+}
+
+void MFAAC_GetDefaults(obs_data_t *settings)
+{
+	obs_data_set_default_int(settings, "bitrate", 128);
+}
+
+void *MFAAC_Create(obs_data_t *settings, obs_encoder_t *encoder)
+{
+	UINT32 bitrate = (UINT32)obs_data_get_int(settings, "bitrate");
+	if (!bitrate) {
+		MF_LOG_ENCODER("AAC", encoder, LOG_ERROR, "Invalid bitrate specified");
+		return nullptr;
+	}
+
+	auto *audio = obs_encoder_audio(encoder);
+	UINT32 channels = (UINT32)audio_output_get_channels(audio);
+	UINT32 sampleRate = audio_output_get_sample_rate(audio);
+	UINT32 bitsPerSample = 16;
+
+	UINT32 recommendedSampleRate = FindBestSamplerateMatch(sampleRate);
+	if (recommendedSampleRate != sampleRate) {
+		MF_LOG_ENCODER("AAC", encoder, LOG_WARNING,
+			       "unsupported sample rate; "
+			       "resampling to best guess '%d' instead of '%d'",
+			       recommendedSampleRate, sampleRate);
+		sampleRate = recommendedSampleRate;
+	}
+
+	UINT32 recommendedBitRate = FindBestBitrateMatch(bitrate);
+	if (recommendedBitRate != bitrate) {
+		MF_LOG_ENCODER("AAC", encoder, LOG_WARNING,
+			       "unsupported bitrate; "
+			       "resampling to best guess '%d' instead of '%d'",
+			       recommendedBitRate, bitrate);
+		bitrate = recommendedBitRate;
+	}
+
+	std::unique_ptr<Encoder> enc(new Encoder(encoder, bitrate, channels, sampleRate, bitsPerSample));
+
+	if (!enc->Initialize())
+		return nullptr;
+
+	return enc.release();
+}
+
+void MFAAC_Destroy(void *data)
+{
+	Encoder *enc = static_cast<Encoder *>(data);
```

Comment:
This is almost fine, though a more elegant way would be something like:

```suggestion
auto encoder = std::unique_ptr<Encoder>{std::exchange(static_cast<Encoder *>(data), nullptr)};
```
Raw `new` and `delete` are bad form in modern C++, and the suggestion makes the ownership transfer more explicit:

* The raw pointer is converted into a `unique_ptr` in the local scope, effectively taking ownership of it (ownership that was released in `MFAAC_Create` originally).
* The `std::exchange` explicitly sets the original pointer to a `nullptr`, so `data` will be `NULL` from `libobs` point of view.
* The destructor of `Encoder` will be called when `MFAAC_Destroy` goes out of scope.

My view is that once the code starts to opt-in to proper ownership management via smart pointers, it needs to keep using the logic for consistency, and this means that after ownership has been relinquished, it needs to be reacquired before anything is done that influences the lifetime of the managed object.